### PR TITLE
Enforce bone hierarchy in SMD exporting

### PR DESCRIPTION
**Pull request by grey-r**
When attempting to export a particular rigged mesh, I kept encountering an error: 
`Blender Source Tools: exporting skeleton_name
- Baking...
- D:\Documents\3D Models\Commissions\xxx\decompiled 0.62\smd2\anims\ragdoll.smd
Traceback (most recent call last):
  File "C:\Users\Grey Ruessler\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\io_scene_valvesource\export_smd.py", line 211, in execute
    self.exportId(context, exportable.get_id())
  File "C:\Users\Grey Ruessler\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\io_scene_valvesource\export_smd.py", line 563, in exportId
    self.files_exported += write_func(id, bake_results, self.sanitiseFilename(export_name), path)
  File "C:\Users\Grey Ruessler\AppData\Roaming\Blender Foundation\Blender\2.80\scripts\addons\io_scene_valvesource\export_smd.py", line 990, in writeSMD
    line += str(self.bone_ids[parent.name])
KeyError: 'ValveBiped.Bip01_Pelvis'`

Based on this KeyError, I came to the conclusion that for one reason or another, ValveBiped.Bip01_Pelvis wasn't iterated through first, and thus not added to the bone_ids array. As iteration is done sequentially, this then implies that the previous inline function wasn't necessarily hitting the root bones first.

My replacement, though less elegant, ensures that the proper hierarchy is followed when the bone tree is flattened; this alleviates the KeyError issue, and should produce more consistent results overall, especially with more than one root bone.

tl;dr
**self.armature.pose.bones isn't guaranteed to be in order**

### CODE REVIEW

**Comment by grey-r** [comment]
Since writing this, I've also fixed one more issue I've found, though I'm not certain what sideffects it may have.
When unapplied modifiers are present, such as subdiv surface or solidify, the bake function was previously discarding important weight data, resulting in a vertex.groups field which contained nothing. Changing the bake object creation to preserve data layers, using the depgraph as already implemented, results in a fix for this.

**Comment by Artfunkel** [comment]
Thanks for the effort, I found my own solutions to both of these issues though.
